src/lava/magma/compiler/builder.py

```diff
@@ -326,13 +353,53 @@ def build(self):
                 csp_ports = self.csp_ports[name]
                 if not isinstance(csp_ports, list):
                     csp_ports = [csp_ports]
-            port = port_cls(pm, csp_ports, p.shape, lt.d_type)
+            port = port_cls(csp_ports, pm, p.shape, lt.d_type)
 
             # Create dynamic PyPort attribute on ProcModel
             setattr(pm, name, port)
             # Create private attribute for port precision
             # setattr(pm, "_" + name + "_p", lt.precision)
 
+        # Initialize RefPorts
+        for name, p in self.ref_ports.items():
+            # Build PyPort
+            lt = self._get_lava_type(name)
+            port_cls = ty.cast(ty.Type[PyRefPort], lt.cls)
+            csp_recv = None
+            csp_send = None
+            if name in self.csp_ports:
+                csp_ports = self.csp_ports[name]
+                csp_recv = csp_ports[0] if isinstance(
+                    csp_ports[0], CspRecvPort) else csp_ports[1]
+                csp_send = csp_ports[0] if isinstance(
+                    csp_ports[0], CspSendPort) else csp_ports[1]
+
+            port = port_cls(csp_send, csp_recv, pm, p.shape, lt.d_type)
+
+            # Create dynamic PyPort attribute on ProcModel
+            setattr(pm, name, port)
+
+        # Initialize VarPorts
+        for name, p in self.var_ports.items():
```

Comment:
There's a lot of code duplication happening here. We should refactor all Port building to make this cleaner.

Reply:
Issue #60 filed